rust/main/hyperlane-base/src/db/error.rs

```diff
@@ -0,0 +1,34 @@
+use std::{io, path::PathBuf};
+
+use hyperlane_core::{ChainCommunicationError, HyperlaneProtocolError};
+
+/// DB Error type
+#[derive(thiserror::Error, Debug)]
+pub enum DbError {
+    /// Rocks DB Error
+    #[error("{0}")]
+    RockError(#[from] rocksdb::Error),
+    #[error("Failed to open {path}, canonicalized as {canonicalized}: {source}")]
+    /// Error opening the database
+    OpeningError {
+        /// Rocksdb error during opening
+        #[source]
+        source: rocksdb::Error,
+        /// Raw database path provided
+        path: PathBuf,
+        /// Parsed path used
+        canonicalized: PathBuf,
+    },
```

Comment:
I wonder if we should hide the error from a particular type of database and expose a generalised error to the user of `HyperlaneDb` trait?

Reply:
`DbError` is generically named, and its members could be anything based on specific implementations - `DbError::RocksError` can only be created from a rocksdb error so it won't be misused. 

`DbError` _is_ made fully generalised when it's converted to `ChainCommunicationError`, and I considered defining the error as an associated type, but usage ends up looking _more_ concrete, e.g. `Arc<dyn HyperlaneDb<DbError = RocksDbError>>`

Due to that I'm not sure it's worth generalising further in the current state, could you give an example?

Reply:
If we just converting into a different error, we don't need to generalise. If we would like to react on the error in a meaningful way, we may want to generalise it.